pygrocy/grocy_api_client.py

```diff
@@ -327,39 +327,65 @@ def __init__(self, base_url, api_key, port: int = DEFAULT_PORT_NUMBER, verify_ss
                 "GROCY-API-KEY": api_key
             }
 
+    def _do_get_request(self, end_url: str):
+        req_url = urljoin(self._base_url, end_url)
+        resp = requests.get(
+            req_url, verify=self._verify_ssl, headers=self._headers)
+        resp.raise_for_status()
+        if len(resp.content) > 0:
+            return resp.json()
+
+    def _do_post_request(self, end_url: str, data: dict):
+        req_url = urljoin(self._base_url, end_url)
+        resp = requests.post(
+            req_url, verify=self._verify_ssl,
+            headers=self._headers,
+            data=data)
+        resp.raise_for_status()
+        if len(resp.content) > 0:
+            return resp.json()
+
+    def _do_put_request(self, end_url: str, data):
+        req_url = urljoin(self._base_url, end_url)
+        up_header = self._headers.copy()
+        up_header['accept'] = '*/*'
+        if isinstance(data, dict):
+            up_header['Content-Type'] = 'application/json'
+            data = json.dumps(data)
+        else:
+            up_header['Content-Type'] = 'application/octet-stream'
+        resp = requests.put(
+            req_url, verify=self._verify_ssl,
+            headers=up_header,
+            data=data)
+        resp.raise_for_status()
+        if len(resp.content) > 0:
+            return resp.json()
+
     def get_stock(self) -> List[CurrentStockResponse]:
-        req_url = urljoin(self._base_url, "stock")
-        resp = requests.get(req_url, verify=self._verify_ssl, headers=self._headers)
-        if resp.status_code != 200 or not resp.text:
-            return
-        parsed_json = resp.json()
+        parsed_json = self._do_get_request("stock")
         return [CurrentStockResponse(response) for response in parsed_json]
 
     def get_volatile_stock(self) -> CurrentVolatilStockResponse:
-        req_url = urljoin(self._base_url, "stock/volatile")
-        resp = requests.get(req_url, verify=self._verify_ssl, headers=self._headers)
-        parsed_json = resp.json()
-        return CurrentVolatilStockResponse(parsed_json)
+        parsed_json = self._do_get_request("stock/volatile")
+        if parsed_json:
+            return CurrentVolatilStockResponse(parsed_json)
 
     def get_product(self, product_id) -> ProductDetailsResponse:
-        req_url = urljoin(urljoin(self._base_url, "stock/products/"), str(product_id))
-        resp = requests.get(req_url, verify=self._verify_ssl, headers=self._headers)
-        if resp.status_code != 200 or not resp.text:
-            return
-        parsed_json = resp.json()
-        return ProductDetailsResponse(parsed_json)
+        url = f"stock/products/{product_id}"
+        parsed_json = self._do_get_request(url)
+        if parsed_json:
```

Comment:
Is there a reason you kept the `if` statement here?
I think the get request will fail anyway if no product exists with the given id, but I may be wrong about that.

Reply:
```if``` is in case of no data, not to test an error.
Because DetailsReponse construcor needs a json.